File: src/diffpath.pl
use strict;
use warnings;
use File::Spec;
use File::Basename;
use File::Find;

# Function to check if a file is executable by user, group, or others
sub is_executable {
    my $file = shift;
    return ( -f $file || -l $file ) && ( ( lstat($file) )[2] & 0111 );
}

# Function to collect all executable files from the given PATH
sub collect_executables {
    my ($path) = @_;
    my %executables;
    my @dirs = split /:/, $path;

    for my $dir (@dirs) {
        next unless -d $dir;
        find(
            {
                follow => 0,
                wanted => sub {

                    # Skip the root directory itself
                    return if $File::Find::name eq $dir;
                    return unless is_executable($_);
                    $executables{$_} = 1;
                }
            },
            $dir
        );
    }
    return \%executables;
}

# Function to calculate the symmetric difference between two hashes
sub symmetric_difference {
    my ( $execs1, $execs2 ) = @_;

    my %unique1 = map { $_ => 1 } grep { !exists $execs2->{$_} } keys %$execs1;
    my %unique2 = map { $_ => 1 } grep { !exists $execs1->{$_} } keys %$execs2;

    my @result = sort ( keys %unique1, keys %unique2 );
    return @result;
}

# Function to print the results in the specified format
sub print_results {
    my ( $result, $unique1, $unique2 ) = @_;

    foreach my $exe (@$result) {
        if ( exists $unique1->{$exe} ) {
            print "$exe\n";
        }
        else {
            print "\t$exe\n";
        }
    }
}

# Main function to execute the script logic
sub main {
    if ( @ARGV != 2 ) {
        die "Usage: $0 PATH1 PATH2\n";
    }
    my ( $path1, $path2 ) = @ARGV;

    # Collect executables from both paths
    my $execs1 = collect_executables($path1);
    my $execs2 = collect_executables($path2);

    # Get the symmetric difference and sort the results
    my @result = symmetric_difference( $execs1, $execs2 );

    # Print the results in the specified format
    print_results( \@result, $execs1, $execs2 );
}

main();

__END__

File: src/diffpath.ts
import fs from "fs";
import path from "path";

// Function to check if a file is executable
function isExecutable(filePath: string): boolean {
  const stats = fs.lstatSync(filePath);
  return (
    stats.isSymbolicLink() || (stats.isFile() && (stats.mode & 0o111) !== 0)
  );
}

// Function to get all executable files in a directory
function getExecutableFiles(paths: string): Set<string> {
  const executables = new Set<string>();
  paths.split(":").forEach((dir) => {
    try {
      const files = fs.readdirSync(dir);
      files.forEach((file) => {
        const filePath = path.join(dir, file);
        try {
          if (isExecutable(filePath)) {
            executables.add(file);
          }
        } catch (err) {
          // Ignore errors for individual files
        }
      });
    } catch (err) {
      // Ignore errors for directories that can't be read
    }
  });
  return executables;
}

// Function to get the symmetric difference between two sets
function symmetricDifferenceList(
  set1: Set<string>,
  set2: Set<string>,
): string[] {
  const difference: string[] = [];
  set1.forEach((value) => {
    if (!set2.has(value)) {
      difference.push(value);
    }
  });
  set2.forEach((value) => {
    if (!set1.has(value)) {
      difference.push(value);
    }
  });
  difference.sort();
  return difference;
}

// Main function to execute the diffpath logic
function main(): void {
  if (process.argv.length !== 4) {
    const absolutePath = process.argv[1];
    const relativePath = path.relative(process.cwd(), absolutePath);
    console.error(`Usage: ${relativePath} PATH1 PATH2`);
    process.exit(1);
  }
  const [, , path1, path2] = process.argv;

  const execSet1 = getExecutableFiles(path1);
  const execSet2 = getExecutableFiles(path2);

  const sortedDiff = symmetricDifferenceList(execSet1, execSet2);

  sortedDiff.forEach((file) => {
    if (execSet1.has(file)) {
      console.log(file);
    } else {
      console.log(`\t${file}`);
    }
  });
}

main();
